DEVO-45 Update Airflow Dev Container README (#10)
tulibraries/airflow-docker-dev-setup: README.md

# Shared Docker Development Setup for Airflow DAGs

This Git repository contains all of the setup needed to run airflow in docker, mount your local DAG repo into the containers that need it, and allow you to develop DAGs hopefully quickly!

## Expected Setup

This repo expects:

* To be a Git submodule residing in the root of your DAG's Git repoitory. 

> * The DAG project directory structure:

Reviewer: @nomadicoder I find this structure to be confusing.  All of the docker files are shown at the top level rather than in the airflow-docker-dev-setup directory.  I think it would be better to take out the whole directory structure out. You could include links to some/all of the repositories instead.

Author: Rewritten

Add this Git repository as a Git submodule to the root of the DAG you are developing.

```
git submodule add git@github.com:tulibraries/airflow-docker-dev-setup docker`
```

Cmmit the changes to .gitmodules and the new `docker` directory that has just been created.

```
git add .gitsubmodules docker
```

In the root of your dags repository, create a file called `local.env`. add the line `DAG_DIR=my_dag_dir` with the actual name of the directory in the root of the repo containing your dag files. Following the example directory tree, the line would be: `DAG_DIR=example_dags` If your dags python code has mport statements, this DAG directory name must match the top level package in those import statements

### <a id="FromGitModule"></a>Setup From Existing DAG Git Repository

If you cloned your DAG from a Git repository and it already has the `airflow-docker-dev-setup` Git submodule, you will need to perform the following commands to complete the dev container installation process.  From your DAG's root directory:

```
	git submodule init
	git submodule update
```

> Then change directories to the docker dev environment `cd docker`

Reviewer: Please use the actual name of the directory:
```suggestion
Then change directories to the docker dev environment `cd airflow-docker-dev-setup`
```

Author: Rewritten

## Usage

> To run Airflow in within a containerized dev environment, go into the `docker` directory: `cd docker `. This directory contains the `docker-compose.yml`, docker configurationss, and a `docker-requirement.txt` for pypi packages you want installed on the container, and a Makefile defining some useful commands.

Reviewer: ```suggestion
To run Airflow in a containerized dev environment, go into the `airflow-docker-dev-setup` directory: `cd airflow-docker-dev-setup `. This directory contains the `docker-compose.yml`, docker configurationss, and a `docker-requirement.txt` for pypi packages you want installed on the container, and a Makefile defining some useful commands.
```

Author: Rewritten

$ docker-compose -p infra ps
```

> **NOTE** You may encounter a `Variale Import Failed` error which will prevent the proper startup of all the Docker containers.  This is a common problem.  See [Variable Import Failed](#VariableImportFailed) below and re-run `make-up`

Reviewer: I can leave this in here, but I have not run into this problem when working in this docker environment.
```suggestion
**NOTE** You may encounter a `Variable Import Failed` error which will prevent the proper startup of all the Docker containers.  This is a common problem.  See [Variable Import Failed](#VariableImportFailed) below and re-run `make-up`
```

Author: Rewrtten

#### Start Bash in Given Docker Instance

Run command shell in Airflow Worker instance:

```
 $ make tty-worker
```

Run command shell in Airflow Webserver instance:

```
$ make tty-webserver
```

Run command shell in Airflow Scheduler instance:

```
$ make tty-scheduler
```

### Start Bash as Root in Given Docker Instance

Run command shell as root in Airflow Worker instance:

```
$ make tty-root-worker
```

Run command shell as root in Airflow Webserver instance:

```
$ make tty-root-webserver
```
Run command shell as root in Airflow Scheduler instance:

```
$ make tty-root-scheduler

> ```

Reviewer: We recently added in both make and lint commands to the Makefile.  Can you add those here too?

Author: Added other Makefile targets.

## <a id="VariableImportFailed"></a>Common Problem: "Variable Import Failed"

> During `make up` you may encounter several erros similar to:

Reviewer: ```suggestion
During `make up` you may encounter several errors similar to:
```

Author: Rewritten

```
   Variable import failed: ProgrammingError('(psycopg2.ProgrammingError) relation "variable" does not exist...
```

> Extend the build target's sleep time in `airflow-docker-dev-setup/Makefile` to allow the MySQL container more time to start up before continuing on to subsequent commands.  Edit `airflow-docker-dev-setup/Makefile`, search for the `build` target and edit the `sleep` command's time from `40` to `120`.

Reviewer: ```suggestion
Extend the build target's sleep time in `airflow-docker-dev-setup/Makefile` to allow the Postgres container more time to start up before continuing on to subsequent commands.  Edit `airflow-docker-dev-setup/Makefile`, search for the `build` target and edit the `sleep` command's time from `20` to `120`.
```

Author: Rewritten
